File: ctci/tmp/chapter_4_trees_graphs/cpp/4.5-validate_bst.cpp
#include "tree.h"

/**
 * Helper.
 */
bool validate_bst_helper(TreeNode *root, int *min, int *max)
{
    if (!root)
        return true;
    if ((min && root->value <= *min) || (max && root->value > *max))
        return false;
    if (!validate_bst_helper(root->left, min, &root->value) ||
        !validate_bst_helper(root->right, &root->value, max))
        return false;
    return true;
}

/**
 * Validates that a binary tree is a binary search tree.
 */
bool validate_bst(TreeNode *root)
{
    return validate_bst_helper(root, nullptr, nullptr);
}

int main()
{
    TreeNode *t1 = new TreeNode(20);
    TreeNode *t2 = t1->add_left(10);
    TreeNode *t3 = t1->add_right(30);
    TreeNode *t4 = t2->add_left(5);
    TreeNode *t5 = t2->add_right(15);
    TreeNode *t6 = t3->add_left(25);
    /* TreeNode *t7 = t3->add_right(27); 0 */
    TreeNode *t7 = t3->add_right(50);

    cout << validate_bst(t1) << endl; // 1

    return 0;
}

File: ctci/tmp/chapter_4_trees_graphs/python/4.5-validate_BST.py
TreeNode = __import__('tree').TreeNode


def validate_bst_helper(root: TreeNode, minimum: int, maximum: int) -> bool:
    """
    Helper.
    """
    if not root:
        return True
    if (minimum and root.value < minimum) or (maximum and root.value > maximum):
        return False
    if not validate_bst_helper(root.left, minimum, root.value) or \
       not validate_bst_helper(root.right, root.value, maximum):
        return False
    return True


def validate_bst(root: TreeNode) -> bool:
    """
    Validates that a binary tree is a binary search tree.
    """
    return validate_bst_helper(root, None, None)


if __name__ == "main":
    n1 = TreeNode(15)
    n2 = n1.add_left(12)
    n3 = n1.add_right(22)
    n4 = n2.add_left(11)
    n5 = n2.add_right(14)
    n6 = n3.add_left(16)
    n7 = n4.add_right(30)
    # n3.add_right(17) # False

    print(validate_bst(n1))  # True
